sw/airborne/modules/dcf/dcf.c

```diff
@@ -0,0 +1,178 @@
+ *
+ */
+
+#include <math.h>
+#include <std.h>
+#include <stdio.h>
+
+#include "modules/dcf/dcf.h"
+#include "subsystems/datalink/datalink.h" // dl_buffer
+#include "subsystems/datalink/telemetry.h"
+#include "subsystems/navigation/common_nav.h"
+#include "autopilot.h"
+
+#if PERIODIC_TELEMETRY
+static void send_dcf(struct transport_tx *trans, struct link_device *dev)
+{
+  pprz_msg_send_DCF(trans, dev, AC_ID, 4 * DCF_MAX_NEIGHBORS, &(tableNei[0][0]), DCF_MAX_NEIGHBORS, error_sigma);
+}
+#endif // PERIODIC TELEMETRY
+
+// Control
+/*! Default gain k for the algorithm */
+#ifndef DCF_GAIN_K
+#define DCF_GAIN_K 10
+#endif
+/*! Default radius for the circumference */
+#ifndef DCF_RADIUS
+#define DCF_RADIUS 80
+#endif
+/*! Default timeout for the neighbors' information */
+#ifndef DCF_TIMEOUT
+#define DCF_TIMEOUT 1500
+#endif
+
+dcf_con dcf_control = {DCF_GAIN_K, DCF_RADIUS, DCF_TIMEOUT, 0};
+
+/*! Default number of neighbors per aircraft */
+#ifndef DCF_MAX_NEIGHBORS
+#define DCF_MAX_NEIGHBORS 4
+#endif
+int16_t tableNei[DCF_MAX_NEIGHBORS][4];
+int16_t error_sigma[DCF_MAX_NEIGHBORS];
+uint32_t last_theta[DCF_MAX_NEIGHBORS];
+uint32_t last_transmision = 0;
+
+void dcf_init(void)
+{
+  for (int i = 0; i < DCF_MAX_NEIGHBORS; i++) {
+    tableNei[i][0] = -1;
+    error_sigma[i] = 0;
+  }
+
+#if PERIODIC_TELEMETRY
+  register_periodic_telemetry(DefaultPeriodic, PPRZ_MSG_ID_DCF, send_dcf);
+#endif
+}
+
+bool distributed_circular(uint8_t wp)
+{
+  float xc = waypoints[wp].x;
+  float yc = waypoints[wp].y;
+  struct EnuCoor_f *p = stateGetPositionEnu_f();
+  float x = p->x;
+  float y = p->y;
+  float u = 0;
+
+  dcf_control.theta = atan2f(y - yc, x - xc);
+
+  uint32_t now = get_sys_time_msec();
+
+  for (int i = 0; i < DCF_MAX_NEIGHBORS; i++) {
+    if (tableNei[i][0] != -1) {
+      uint32_t timeout = now - last_theta[i];
+      if (timeout > dcf_control.timeout) {
+        tableNei[i][3] = dcf_control.timeout;
+      } else {
+        tableNei[i][3] = (uint16_t)timeout;
+
+        float e = dcf_control.theta - (tableNei[i][1] + (tableNei[i][2])) * M_PI / 1800.0;
+        if (e > M_PI) {
+          e -= 2 * M_PI;
+        } else if (e <= -M_PI) {
+          e += 2 * M_PI;
+        }
+
+        u += e;
+        error_sigma[i] = (uint16_t)(e * 1800.0 / M_PI);
+      }
+    }
+  }
+
+  u *= -gvf_control.s*dcf_control.k;
+
+  gvf_ellipse_XY(xc, yc, dcf_control.radius + u, dcf_control.radius + u, 0);
+
+  if ((now - last_transmision > 200) && (autopilot_get_mode() == AP_MODE_AUTO2)) {
+    send_theta_to_nei();
+    last_transmision = now;
+  }
+
+  return true;
+}
+
+void send_theta_to_nei(void)
+{
+  struct pprzlink_msg msg;
+
+  for (int i = 0; i < DCF_MAX_NEIGHBORS; i++)
+    if (tableNei[i][0] != -1) {
+      msg.trans = &(DefaultChannel).trans_tx;
+      msg.dev = &(DefaultDevice).device;
+      msg.sender_id = AC_ID;
+      msg.receiver_id = tableNei[i][0];
+      msg.component_id = 0;
+      pprzlink_msg_send_DCF_THETA(&msg, &(dcf_control.theta));
```

Comment:
If you are sending the same data to all the aircraft, why send it multiple time? Why not just send it once with the msg.receiver_id as a broadcast id like 255 or 0. Actually, you don't check the id in the parseThetaTable atm so it seems like the messages will just be received multiple times. 

I am not totally aware how the new messages are handled, @gautierhattenberger does the pprzlink decoder only propagate messages that are designated to you or do you need to check that yourself?

Reply:
It is not to all the aircraft, but to a subset of them (the neighbors of the aircraft). You can have ten aircraft flying in total but a particular aircraft only needs to interact with a subset of them. The algorithm is distributed in the sense that not everybody needs to share information with everybody.

Yes, in fact for XBee radios, the filtering of the message happens at the radio chip, since the ID of the aircraft is the same as the ID for the radio.